pkg/etcdcli/etcdcli.go

```diff
@@ -101,27 +83,73 @@ func (g *etcdClientGetter) getEtcdClient() (*clientv3.Client, error) {
 		}
 		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", bootstrapIP))
 	}
+	return etcdEndpoints, nil
+}
+
+type etcdClient struct {
+	EtcdEndpointsGetter
+
+	// TODO: need to find a way to make controllers specify their own even recorder
+	//       instead of generic event recorder, so the events can be back traced to
+	//      the controller calling the client methods
+	eventRecorder events.Recorder
+
+	clientLock          sync.Mutex
+	lastClientConfigKey []string
+	cachedClient        *clientv3.Client
+}
 
-	g.clientLock.Lock()
-	defer g.clientLock.Unlock()
-	// TODO check if the connection is already closed
-	if reflect.DeepEqual(g.lastClientConfigKey, etcdEndpoints) {
-		return g.cachedClient, nil
+func NewEtcdClient(kubeInformers v1helpers.KubeInformersForNamespaces, networkInformer configv1informers.NetworkInformer, eventRecorder events.Recorder) EtcdClient {
+	return &etcdClient{
+		EtcdEndpointsGetter: &etcdClientGetter{
+			nodeLister:            kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+			endpointsLister:       kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Endpoints().Lister(),
+			networkLister:         networkInformer.Lister(),
+			nodeListerSynced:      kubeInformers.InformersFor("").Core().V1().Nodes().Informer().HasSynced,
+			endpointsListerSynced: kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Endpoints().Informer().HasSynced,
+			networkListerSynced:   networkInformer.Informer().HasSynced,
+		},
+		eventRecorder: eventRecorder.WithComponentSuffix("etcd-client"),
+	}
+}
+
+// getCachedEtcdClient may return a cached client.  When a new client is needed, the previous client is closed.
+// The caller should not close the client or future calls may fail.
+func (e *etcdClient) getCachedEtcdClient() (*clientv3.Client, error) {
+	etcdEndpoints, err := e.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	e.clientLock.Lock()
+	defer e.clientLock.Unlock()
+	// check if the connection is already closed
+	if reflect.DeepEqual(e.lastClientConfigKey, etcdEndpoints) &&
+		e.cachedClient.ActiveConnection().GetState() == connectivity.Ready {
+		return e.cachedClient, nil
 	}
 
 	c, err := getEtcdClient(etcdEndpoints)
 	if err != nil {
 		return nil, err
 	}
-	if g.cachedClient != nil {
-		if err := g.cachedClient.Close(); err != nil {
+	if e.cachedClient != nil {
+		if err := e.cachedClient.Close(); err != nil {
 			utilruntime.HandleError(err)
 		}
 	}
-	g.cachedClient = c
-	g.lastClientConfigKey = etcdEndpoints
+	e.cachedClient = c
+	e.lastClientConfigKey = etcdEndpoints
 
-	return g.cachedClient, nil
+	// set endpoints to existing members
```

Comment:
This will avoid calls to bootstrap endpoint even if it is present during original client creation.

Reply:
will fix https://github.com/openshift/okd/issues/139